plugins/ltac2/tac2intern.ml

```diff
@@ -1166,6 +1166,14 @@ and intern_constructor env loc kn args = match kn with
       (GTacOpn (kn, args), ans)
   else
     error_nargs_mismatch ?loc kn nargs (List.length args)
+| Tuple 0 when not (List.is_empty args) ->
+  (* parsing [() bla] produces [CTacApp (Tuple 0, [bla])] but parsing
+     [((), ()) bla] produces [CTacApp (CTacApp (Tuple 2, [(); ()]), [bla])]
+     so we only need to produce a sensible error for [Tuple 0] *)
+  let t = GTypRef (Tuple 0, []) in
+  CErrors.user_err ?loc Pp.(
+      str "This expression has type" ++ spc () ++ pr_glbtype env t ++
+      spc () ++ str "and is not a function")
 | Tuple n ->
   assert (Int.equal n (List.length args));
```

Comment:
Why not move the above code here? We're already performing an equality check here, and where clauses are not robust.